### ISSUE TITLE
::Val{x} sometimes performs significantly worse than ::Type{Val{x}}

### ISSUE THREAD

**Issue by benninkrs** [JuliaLang/julia#37675]
Ever since [Base functions were changed to take `Val(x)` instead of `Val{x}`](https://docs.julialang.org/en/v0.7/NEWS/#Library-improvements-1) I have had the impression that `Val{x}` had been relegated to legacy . However, I have found that there are cases where using `Val(x)` leads to significantly worse performance than using `Val{x}`. Below is a MWE:
```
function pass_value(A::AbstractArray{Float64,N}, p::Dims{N}) where {N}
    loop!(A, Val(p))
end

function pass_type(A::AbstractArray{Float64,N}, p::Dims{N}) where {N}
    loop!(A, Val{p})
end

function loop!(A::AbstractArray, ::Val{p}) where {p}
    for I in CartesianIndices(A)
        I_ = map(j->I[j], p)
        A[I_...] += 1.0
    end
end

function loop!(A::AbstractArray, ::Type{Val{p}}) where {p}
    for I in CartesianIndices(A)
        I_ = map(j->I[j], p)
        A[I_...] += 1.0
    end
end

julia> using BenchmarkTools

julia> A = rand(4,4,4,4,4);

julia> @btime pass_value($A, (5,4,3,2,1))
  4.063 μs (1 allocation: 48 bytes)

julia> @btime pass_type($A, (5,4,3,2,1))
  974.917 ns (1 allocation: 48 bytes)
```
I *think* that within larger projects I have also observed a performance gap with simpler type parameters (integers instead of tuples), but so far I haven't been able to reproduce it in a simple example.

**Comment by eschnett** [comment]
Can you post details of your system configuration, i.e. Julia version, CPU type, OS, etc.?

I'm using
```Julia
julia> versioninfo()
Julia Version 1.6.0-DEV.952
Commit 4ef8313c94 (2020-09-17 11:39 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.6.0)
  CPU: Intel(R) Core(TM) i7-8850H CPU @ 2.60GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
```
and I can't reproduce this. I get no significant difference:
```Julia
julia> @btime pass_value($A, (5,4,3,2,1))
  1.061 μs (1 allocation: 48 bytes)

julia> @btime pass_type($A, (5,4,3,2,1))
  1.068 μs (1 allocation: 48 bytes)
```

**Comment by benninkrs** [comment]
julia> versioninfo()
Julia Version 1.5.0
Commit 96786e22cc (2020-08-01 23:44 UTC)
Platform Info:
  OS: Windows (x86_64-w64-mingw32)
  CPU: Intel(R) Core(TM) i7-8650U CPU @ 1.90GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-9.0.1 (ORCJIT, skylake)
Environment:
  JULIA_NUM_THREADS = 4

**Comment by mcabbott** [comment]
This disappears on 1.6 for me:
```
julia> @btime pass_value($A, (5,4,3,2,1))
  1.156 μs (1 allocation: 48 bytes)

julia> @btime pass_type($A, (5,4,3,2,1))
  1.168 μs (1 allocation: 48 bytes)

julia> VERSION
v"1.6.0-DEV.927"
```
vs:
```
  4.912 μs (1 allocation: 48 bytes)
  1.148 μs (1 allocation: 48 bytes)
v"1.5.0"
```

```
  7.199 μs (1 allocation: 48 bytes)
  1.511 μs (1 allocation: 48 bytes)
v"1.4.2"
```

```
  8.858 μs (1 allocation: 48 bytes)
  4.120 μs (1 allocation: 48 bytes)
v"1.3.1"
```

**Comment by dkarrasch** [comment]
I can reproduce the issue on v1.5.1 (on MacOS), but the issue seems resolved in v1.6. 🎉

**Comment by benninkrs** [comment]
Great! It seems this issue can be closed then.
